### PR TITLE
Changing interval frequency field to use go standard time format

### PR DESCRIPTION
**Pull request by xmlviking**
Issue #1246  - Adjust Interval frequency so that uses Go standard time format.
Should allow miliseconds, microseconds, as well as nonoseconds.

Signed-off-by: xmlviking <ecotter@gmail.com>

### CODE REVIEW

File: cmd/config-seed/res/properties/go.mod [hunk 1]
```diff
@@ -1 +1,3 @@
 module github.com/edgexfoundry/edgex-go/cmd/config-seed/res/properties
+
+go 1.12
```
Comment (brandonforster) [inline cmd/config-seed/res/properties/go.mod]: Why do you need Go 1.12?

File: internal/support/scheduler/const.go [hunk 2]
```diff
@@ -28,7 +28,6 @@ const (
 	VALUE                = "value"
 	UNLOCKED             = "UNLOCKED"
 	ENABLED              = "ENABLED"
-	SCHEDULER_TIMELAYOUT = "20060102T150405" /** Scheduler **/
 	TIMELAYOUT           = "20060102T150405" /** Scheduler **/
```
Comment (brandonforster) [inline internal/support/scheduler/const.go]: Is this comment still valid?

File: internal/support/scheduler/schedulecontext.go [hunk 3]
```diff
@@ -70,7 +67,11 @@ func (sc *IntervalContext) Reset(interval models.Interval) {
 	//frequency and next time
 	nowBenchmark := time.Now().Unix()
 	if !sc.Interval.RunOnce {
-		sc.Frequency = parseFrequency(sc.Interval.Frequency)
+		frequency, err := parseFrequency(sc.Interval.Frequency)
+		if err != nil {
+			LoggingClient.Error("parse frequency error, the original frequency string is : " + sc.Interval.Frequency)
```
Comment (brandonforster) [inline internal/support/scheduler/schedulecontext.go]: Right now, you're eating the error message.

File: internal/support/scheduler/utils.go [hunk 4]
```diff
@@ -48,6 +49,30 @@ func msToTime(ms string) (time.Time, error) {
...
+// Hour = 60 * Minute
+func parseFrequency(durationStr string) (time.Duration, error) {
+
+	// default
+	var durationDefault = 24 * time.Hour
+
+	// check Frequency
+	timeDuration, err := time.ParseDuration(durationStr)
+	if err != nil {
+		LoggingClient.Error("parse frequency error, the original time string is : " + durationStr)
```
Comment (brandonforster) [inline internal/support/scheduler/utils.go]: This is the same output that your error handling code is going to have. Your code will output the same error alert twice, both without any real insight into what exactly is unparsable.

File: internal/support/scheduler/utils.go [hunk 5]
```diff
@@ -14,25 +14,26 @@
...
 
-const (
-	frequencyPattern = `^P(\d+Y)?(\d+M)?(\d+D)?(T(\d+H)?(\d+M)?(\d+S)?)?$`
-)
-
+// Frequency indicates how often the specific resource needs to be polled.
+// It represents as a duration string.
+// The format of this field is to be an unsigned integer followed by a unit which may be "ms", "s", "m" or "h"
+// representing milliseconds, seconds, minutes or hours. Eg, "100ms", "24h"
 func isFrequencyValid(frequency string) bool {
```
Comment (brandonforster) [inline internal/support/scheduler/utils.go]: This function is simple enough that I think we should refactor it out. It's not adding a whole lot of value.

File: internal/support/scheduler/const.go (hunk 2 above)
Comment (xmlviking) [reply internal/support/scheduler/const.go]: Yes :) that is ISO 8601 in general

Comment (brandonforster) [reply internal/support/scheduler/const.go]: What is "scheduler" telling us because I don't understand the point of the comment. These are all scheduler constants.

File: internal/support/scheduler/utils.go (hunk 5 above)
Comment (xmlviking) [reply internal/support/scheduler/utils.go]: I left it because it's being used by other potential functions.

Comment (brandonforster) [reply internal/support/scheduler/utils.go]: Then remove it there too.

Comment (xmlviking) [reply internal/support/scheduler/utils.go]: after our discussion I will be refactoring this util function out and just modify the code to use the parseDuration func :)

File: internal/support/scheduler/utils.go (hunk 4 above)
Comment (xmlviking) [reply internal/support/scheduler/utils.go]: fixed

**Comment by mhall119** [comment]
Is this going to cause configuration files made for EdgeX 1.0.x break? Is breaking backwards compatibility for configuration files considered a major-version-changing break?

File: internal/support/scheduler/schedulecontext.go (hunk 3 above)
Comment (xmlviking) [reply internal/support/scheduler/schedulecontext.go]: fixed . logging error message as well.

**Comment by xmlviking** [comment]
> Is this going to cause configuration files made for EdgeX 1.0.x break? Is breaking backwards compatibility for configuration files considered a major-version-changing break?

Michael as it sits right now this would break backward compatibility.  The request was to update scheduler interval's to use more prevalent go time format for frequency.

File: internal/support/scheduler/const.go (hunk 2 above)
Comment (xmlviking) [reply internal/support/scheduler/const.go]: removed the comment :) . fixed

File: cmd/config-seed/res/properties/go.mod (hunk 1 above)
Comment (xmlviking) [reply cmd/config-seed/res/properties/go.mod]: removed!  :)

**Comment by anonymouse64** [comment]
Is it possible to implement this in a backward compatible way? I'm not sure what the syntactical relationship between ISO-8601 and RFC3339Nano is, but if there's no overlap between a valid ISO-8601 date and a RFC3339Nano date, then we could implement "smart parsing" which supports both formats automatically. 

If there is overlap between the two formats, then perhaps for Fuji to not break backwards compatibility we add a new `FrequencyFormat` key which defaults to `"ISO-8601"` but also supports a `"RFC3339Nano"` value, then for services wishing to use RFC3339Nano, they could add that key and get the newer behavior.

**Comment by xmlviking** [comment]
> Is it possible to implement this in a backward compatible way? I'm not sure what the syntactical relationship between ISO-8601 and RFC3339Nano is, but if there's no overlap between a valid ISO-8601 date and a RFC3339Nano date, then we could implement "smart parsing" which supports both formats automatically.
> 
> If there is overlap between the two formats, then perhaps for Fuji to not break backwards compatibility we add a new `FrequencyFormat` key which defaults to `"ISO-8601"` but also supports a `"RFC3339Nano"` value, then for services wishing to use RFC3339Nano, they could add that key and get the newer behavior.

Problem is the existing Intervals in an existing database.  Backward compatibility.

**Comment by anonymouse64** [comment]
> Problem is the existing Intervals in an existing database. Backward compatibility.

Well I'm not too familiar with the persistence implementation, but if we are storing the format specified directly then all that would happen is logic working with that would learn to also check for the `FrequencyFormat` specifier, if it was absent or empty it would work the same, otherwise parsing would continue based on what format was selected with that key.

If the Interval is converted to a time.Duration, then it probably would just work? If the format is deserialized out of the database into a time.Duration (i.e. just used directly) then no parsing would be necessary and it wouldn't matter if a Fuji or Edinburgh client wrote the entry into the database as both the ISO-8601 and RFC3339Nano would be serialized into the same time.Duration. If the format is yet some other different serialization (i.e. not the string from configuration.toml and also not a time.Duration), then I assume it could still work the same as described initially, checking if a `FrequencyFormat` key exists and basing deserialization logic on that. 

Finally, I thought that @tsconn23 said we weren't going to support direct database migrations between releases even if those releases were backwards compatible?

**Comment by tsconn23** [comment]
> Finally, I thought that @tsconn23 said we weren't going to support direct database migrations between releases even if those releases were backwards compatible?

@anonymouse64 You might be thinking of what I said in the fourth bullet point on [this comment](https://github.com/edgexfoundry/edgex-go/issues/1424#issuecomment-501488802). The gist is that transitory data like events and readings should be drained before doing a minor version upgrade. All other data should be considered long-lived for the life of the major version as we don't want someone to have to re-provision their whole environment going from Edinburgh --> Fuji

**Comment by anonymouse64** [comment]
@xmlviking can you speak to my questions about how intervals get serialized / deserialized into databases?

> All other data should be considered long-lived for the life of the major version as we don't want someone to have to re-provision their whole environment going from Edinburgh --> Fuji

Understood, thanks for clarifying

**Comment by xmlviking** [comment]
Ian,

The issue with backward compatibility relates to what you mentioned.
Existing installations with Scheduler "Intervals" with the previous version
of the frequency format.  The actually date time is fine and is ISO 8601 so
we will be sticking with that for now.
Again to ensure were not breaking anyone's existing EdgeX environment were
holding off on the frequency change for now and wait till Fuji.

Eric

**Comment by anonymouse64** [comment]
Thanks @xmlviking, unfortunately I didn't quite understand your comment as it doesn't seem to doesn't answer my question about how the scheduler intervals are serialized to a database like Mongodb.

In the meantime, I took the liberty of looking at the code myself and it seems that we serialize intervals with the ISO-8601 string exactly, which means that we could still implement this change in a backwards compatible manner by doing the following:

- add a new FrequencyFormat string to the Interval spec (adding new strings doesn't break backwards compatibility for semver)
- if the FrequencyFormat key is defined for a given Interval, then use that value to deserialize the Frequency string into a time.Duration
- if the FrequencyFormat key is missing or not defined, then assume the format is ISO-8601 and deserialize appropriately

None of these changes would break an existing installation, though note I am assuming that mongo/redis can handle a new key added to a struct when serializing it.

**Comment by xmlviking** [comment]
@anonymouse64 - Ian could you please re-review this change request.  We decided that not changing the schema per say to include the format would not be required at this point.  The code simply supports both formats legacy format as well as the goLang style.  Changes have been made and implemented in the go-mod-contracts to support this bifurcation in the acceptable frequency format.  TY :)

**Comment by anonymouse64** [comment]
> We decided that not changing the schema per say to include the format would not be required at this point

I'm a little confused by this statement, did you change the schema or not? And is a format now included or not?

**Comment by xmlviking** [comment]
> > We decided that not changing the schema per say to include the format would not be required at this point
> 
> I'm a little confused by this statement, did you change the schema or not? And is a format now included or not?

No change in the schema.   No format field.   Were just accepting both old and new formats. It's the least invasion option.

**Comment by xmlviking** [comment]
> > We decided that not changing the schema per say to include the format would not be required at this point
> 
> I'm a little confused by this statement, did you change the schema or not? And is a format now included or not?
Ian,

Based on my analysis of the required areas in EdgeX-go I felt the path of least resistance might be a better approach.   Rather than include a new schema change to the type Interval I chose to enhance the go-mod-contracts serialize and deserialize functions.  I had to do this anyhow regardless of the approach I took.   I actually started with the full change you suggested but found I did not have to do that if I accounted for both the existing frequency time format and the proposed Go Lang standard way.   

In summary I made changes to the go-mod-core contracts interval in the frequency validation section.  Additionally, I updated the validation for the interval frequency in the scheduler as well.   I have not included a schema change for any type persistence.  Nor have I included a new format field

**Comment by anonymouse64** [comment]
Great, thanks for the in depth response Eric. 

I personally won't be able to review this in depth this week, but it has my general +1 for the approach. If it's still open next week I will review, but feel free to merge without my approval this week

**Comment by xmlviking** [comment]
> Great, thanks for the in depth response Eric.
> 
> I personally won't be able to review this in depth this week, but it has my general +1 for the approach. If it's still open next week I will review, but feel free to merge without my approval this week

still open requires rebase but requires your review Ian.   Let me know if you can't get around to this one and I can tap someone else.  Thanks :)

**Comment by xmlviking** [comment]
@brandonforster can I assume your changes have been integrated satisfactory for this PR?  I just don't want to assume they have been.

File: internal/support/scheduler/schedulercontext_test.go [hunk 6]
```diff
@@ -182,26 +182,62 @@ func TestIsComplete(t *testing.T) {
...
+	if err != nil {
+		t.Errorf(TestUnexpectedMsgFormatStrForInt64Val, duration.Nanoseconds(), 50)
+	}
+	if duration.Nanoseconds() != int64(50) {
+		t.Errorf(TestUnexpectedMsgFormatStrForInt64Val, duration.Nanoseconds(), 50)
+	}
+}
+
+// Note Time.Duration does not support milliseconds, or microseconds directly.
+func TestParseMicrosecondsFrequency(t *testing.T) {
```
Comment (akramtexas) [inline internal/support/scheduler/schedulercontext_test.go]: - Making a recommendation for future unit-testing organization. While these tests are thorough, they could benefit from conciseness by the use of a `table-driven` format (it really helps cuts through repetitive testing steps), and an example of which would be the unit tests in [operation_test.go]( https://github.com/edgexfoundry/edgex-go/blob/master/internal/system/agent/executor/operation_test.go)
- The PR as it stands LGTM 👍 
- Approving the PR, pending building checks following the `rebase`.

**Comment by xmlviking** [comment]
recheck
